### DISCUSSION TITLE
Error Handling & Combining Diarization with Timestamps

### DISCUSSION THREAD

**Discussion by CarmAero-1** [deepgram/community#491] [General help]
Hi there, 
We're interested in using Nova 2 for our system's transcription services. However, to ensure the best result, we might need to adjust some of our backend processes to match your offering. 
We have a few questions (JIC it makes a difference, the following are performed in a concurrent transcription application environment), specifically: 
1.	If Nova 2 were to encounter any parts of the audio that are unintelligible or inaudible, how does it flag or present these sections of data?

2.	Is it possible to use diarization in combination with time coding/stamping? If so, how can we do this?

3.	We read through your Q&A "Timecodes in Deepgram results #477" and found it very informative. However, we need some clarification for our particular use case. When employing the speaker diarization, we need to provide the timestamp/timecode whenever there is a change of speaker at the beginning of each section or paragraph. How can we do this? 

Thanks in advance for your help.

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jjmaldonis** [comment]
Hey @CarmAero-1 good questions.

1. When an inaudible or unintelligible part of the audio is encountered, no words will be transcribed. This will work for music, extremely poor audio quality, and silence. To identify these parts of the audio, you can carefully inspect the word-level timestamps in the JSON response: the JSON response contains information about every word transcribed, including when the word started and stopped. If you look for long gaps between word-end and next-word-start, you will be able to identify what parts of the audio are inaudible.

2. Yes, Deepgram supports diarization in combination with word timings out of the box. Simply add `&diarize=true` as a query parameter, and each word will be augmented with a speaker ID. You may also find the query parameters `&paragraph=true` and `&utterances=true` useful.
    a. https://developers.deepgram.com/docs/paragraphs
    b. https://developers.deepgram.com/docs/utterances
3. Check out the paragraphs and utterances features above. Will these fit your use case? Each sentence and utterance has a start time and end time associated with the sentence/utterance.

**Comment by CarmAero-1** [comment]
Jason, thanks for your quick response!

 

The information you provided is indeed suitable for our use case, during our reviews, while we were aware of the utterances section, we totally missed the information regarding paragraphs.

 

A sample of the desired outcome we desire is as follows for “Timestamped Clean Verbatim”:

 

“ 00:00:02] Interviewer: Okay, it’s recording. I’m going to ask you a few questions, answer them briefly or you can add something more if you wish.

[00:00:11] Interviewee: Okay.

[00:00:12] Interviewer: First of all, how old are you?

[00:00:14] Interviewee: I’m 26 at the moment.

[00:00:16] Interviewer: What comes to your mind when you hear the term ‘video games’? “

 

For our “Timestamped Full Verbatim”, we will attempt to parallel the suggestions provided in connection with “ Timecode/Timestamp big chunks. #133” since with this option we will need timestamping  at the beginning of the audio recording and every 2 minutes thereafter.

 

I’m assuming that to achieve the formatting described above, we will likely have to perform post-processing of Nova-2 output (in case we missed something else in your documentation).

 

Lastly, in our Python-based transcription application, we currently use 'Language-Tool-Python' and NLP' for spell-checking, punctuation, grammatical functions, and contextual recognition. 

 

We’re curious to know if such a level of sophistication is necessary for an industry-standard end product. Additionally, if it is required, we would like to know if you foresee any integration issues with your service, with an eye towards ramping up to include hosting on premises.

 

Thanks for your assistance.

**Comment by jjmaldonis** [comment]
Glad the paragraphs feature is what you're looking for!

You may also be interested in our `smart_format` [feature](https://developers.deepgram.com/docs/smart-format). This feature formats transcripts by applying grammer and punctuation. It's likely to perform similarly to your Language Tool Python tool, and it may be worth doing a side-by-side to compare results.

For contextual recognition, Deepgram offers various AI-powered speech analytics features including intent recognition and summarization. These features are in Beta, and you can get a glimpse of them [here](https://deepgram.com/solutions/speech-analytics). If you want to describe your use case in more detail, I can help identify which feature will work best for your product.

I'd also like to mention our `multichannel` [feature](https://developers.deepgram.com/docs/multichannel-vs-diarization), which works well with our `diarization` feature. If your audio has multichannels (e.g. a phone call or video conference with different attendees) and one person is speaking on each channel, you can get perfect diarization by using setting `multichannel=true`.

Finally, I wrote a Python script that formats Deepgram's JSON response in a format similar to what you're looking for.

Here is the curl request I used to save the JSON ([attached here](https://github.com/deepgram/community/files/13677817/example.json)):
```
curl -X POST -H "Authorization: Token $DEEPGRAM_API_KEY" -T /path/to/en_NatGen_CallCenter_BethTom_CancelPhonePlan.wav "https://api.deepgram.com/v1/listen?multichannel=true&diarize=true&paragraphs=true&smart_format=true"
```

Also [attached](https://github.com/deepgram/community/files/13677851/audio.zip) is the audio file used for this example.

Below is the code, followed by the output:

```
import json


def add_channel_to_paragraph(p: dict, channel: int) -> dict:
    p["channel"] = channel
    return p


def format_time(seconds: float) -> str:
    hours = seconds // 3600
    seconds = seconds - (hours * 3600)
    minutes = seconds // 60
    seconds = seconds - (minutes * 60)
    return f"{int(hours):02d}:{int(minutes):02d}:{int(seconds):02d}"


def main():
    with open("/path/to/example.json", "r") as f:
        data = json.load(f)

    # Combine the paragraphs across channels and sort them by the time they were spoken
    nchannels = len(data["results"]["channels"])
    joined_paragraphs = [
        add_channel_to_paragraph(p, i)
        for i, channel in enumerate(range(nchannels))
        for p in data["results"]["channels"][channel]["alternatives"][0]["paragraphs"][
            "paragraphs"
        ]
    ]
    joined_paragraphs.sort(key=lambda p: p["start"])

    # If there are multiple speakers in each channel, we need to give them a unique ID
    speaker_id = {}  # key = (channel_id, speaker_id)
    for p in joined_paragraphs:
        key = (p["channel"], p["speaker"])
        if key not in speaker_id:
            speaker_id[key] = len(speaker_id.keys())

    # Print the paragraphs spoken by each speaker
    for p in joined_paragraphs:
        start = format_time(p["start"])
        speaker = speaker_id[(p["channel"], p["speaker"])]
        sentences = " ".join([s["text"] for s in p["sentences"]])
        print(f"{start} Speaker {speaker}: {sentences}")
        print()


if __name__ == "__main__":
    main()
```
Output:

```
00:00:00 Speaker 0: Hi. Thank you for calling Premier phone services. This call may be recorded for quality and training purposes. My name is Tom and I'll be assisting. You.

00:00:07 Speaker 0: How are you today?

00:00:08 Speaker 1: I'm good. Thank you.

00:00:10 Speaker 0: Alright. That's good. Can I please have your name?

00:00:12 Speaker 1: My name is Beth.

00:00:14 Speaker 0: Okay. Beth. What is your last name?

00:00:17 Speaker 1: Idle.

00:00:19 Speaker 0: Could you spell that for me?

00:00:21 Speaker 1: Yeah. It's just like it sounds, IDLE.

00:00:25 Speaker 0: Okay. I'm not showing Beth idol in our system, but there is an Elizabeth.

00:00:31 Speaker 1: Oh, yeah. That's my full first name. Yeah. That's me.

00:00:35 Speaker 0: Okay, Miss S. What can I do for you today?

00:00:37 Speaker 1: Erin need some help with my phone plan?

00:00:40 Speaker 0: Sure. I can help you help you with that. Can you tell me what plan you currently have?

00:00:45 Speaker 1: I have the Platinum plan.

00:00:48 Speaker 0: Okay. Platinum. How many people do you have on your plan right now?

00:00:54 Speaker 1: That 4 people, myself employed.

00:00:59 Speaker 0: Alright. And how can I help you out with you plan today, ma'am?

00:01:03 Speaker 1: You can help me cancel it. I don't wanna continue my service with you guys.

00:01:08 Speaker 0: Alright. Well, I'm very sorry to hear that, miss Idle. I can assist you and canceling your plan today if. I may ask, do you mind explaining why you wish to terminate your service plan?

00:01:19 Speaker 1: I mean, I don't really feel like I owe you an explanation for why I wanna cancel.

00:01:24 Speaker 0: No. Of course, I will just help us better serve our customers in future by understanding the nature of just dissatisfaction. We can make improvements to our products.

00:01:33 Speaker 1: You mean that you can keep from going bankrupt.

00:01:37 Speaker 0: Yeah. So many words, we want customers to be Saas, their service continue the business with us. Like I said, you don't have provide me an explanation, but it would be very useful to us to add that insight into what the issue is for plan and the reasons why you're canceling it.

00:01:52 Speaker 1: Oh, yeah. I I know I was rude. I'm sorry. I'm just kind of upset because I just got this phone, like, 2 weeks ago, and I'm already having to cancel. So basically, my issue is with the graphics.

00:02:05 Speaker 0: K. What seems to be the issue with the graphics? Precisely.

00:02:09 Speaker 1: Well, they just kind of suck. I mean, this phone is, like, brand new, and the quality of some of the games that I play is just awful. Like, I can't hardly see what's even on the screen. The things will go fuzzy and blurry, sometimes it won't load images. Like, it'll crash if I try to to load an image or play a game that's super high res.

00:02:31 Speaker 1: The... And the emojis in the text messages. It won't load those, much less any pictures or gifts or anything that people send to me. And whenever I try to. Stream videos that are a higher resolution, like, on Youtube, it just completely crashes the app.

00:02:48 Speaker 1: It just can't handle that kind of resolution. So you can even take pictures on this thing either because they're so out of focus. It's just super disappointing, especially since I got the platinum plan for my whole family. I'm spending over 50 bucks a month. And it's just totally not worth it when everyone is complaining to me about how bad the video and pictures are.

00:03:08 Speaker 1: And I know I can get a better phone for about the same price somewhere else. I guess, that's what I should've done in the first place, but my husband really recommended you guys, so I listen to her, I I guess I was just really expecting more.

00:03:21 Speaker 0: K. I understand that you're just putting about the quality graphics on the phone.

00:03:30 Speaker 1: Yeah. It's not just the phone either. My dad has 1 of the player tablets that he uses for Zoom meetings for his work. And he says everyone's always was complaining that his video freezes or he breaks up and they can't understand them. So he's unable to use the audio because that's the only way that the tablet will work.

00:03:47 Speaker 1: Which sucks because during Covid and everything, I think it's still important to try to feel like you're still in that face to face environment, you know? So you can't even do that with this thing.

00:03:58 Speaker 0: Okay. And just for reference, can you tell me the names of all the devices you have your plan? That way, we can... See if there's a problem with the manufacturing level?

00:04:06 Speaker 1: Yeah. I have 1 Strider x s not in 9. That one's from 20 20. I've got 2 of the fidget, x xs r sevens from 20 19. And the player tablet is a v 02:09 0.

00:04:26 Speaker 1: That's dated 20 15.

00:04:31 Speaker 0: Okay. I would guess that the player tablet might be lacking graphics features because it's a bit of an older device compared to the others The actually are going to stop making those parts that's part that model here. And maybe while you're having problems with a particular device.

00:04:45 Speaker 1: Oh, yeah. I guess that would make sense.

00:04:48 Speaker 0: Could I maybe see if we can exchange your device to a new model with the player tablet that might serve your needs better.

00:04:55 Speaker 1: No. I don't think so. I'm just afraid it would have the same problems. So I'm just gonna cancel.

00:05:02 Speaker 0: I can completely understand concerns. Again, I'm I'm very sorry you experienced this frustration.

00:05:06 Speaker 1: Well I mean, it's not your fault. So like I said, I'm sorry if I'm being rude. It's just that I definitely haven't gotten my money's worth with this plan. And I just need to get something else that meets my needs and those with my family. You know, these days, we've realized rely so much in our devices to even do our jobs.

00:05:25 Speaker 1: So they need to meet basic standards as far as the graphics and and video and stuff as long as well as supporting streaming because we're all watching movies at home. Does that make sense?

00:05:36 Speaker 0: Yeah. That make sense. Well, I can get through your order for you today Again, is there any chance you might be interested in another device we could offer you?

00:05:43 Speaker 1: No. I don't think so. I'm I'm just just gonna try another company.

00:05:47 Speaker 0: Alright. Well, we're very sorry to lose your business. To start can I have your account number?

00:05:52 Speaker 1: Yeah. It's 325125675914KI

00:06:05 Speaker 0: And can you confirm your password for me?

00:06:08 Speaker 1: It's busy as heck x 31, the BAH and the x are all capitalized.

00:06:18 Speaker 0: Alright. Okay. I got that. And your phone number is still 4448149913I

00:06:29 Speaker 1: mean, it is today, tomorrow, it'll be something different after I get a new phone.

00:06:34 Speaker 0: Oh, will, would allow you to transfer your phone number even if you cancel. Choose a new device with another company. You can still take the phone number with you.

00:06:43 Speaker 1: Oh, okay. That's how awesome.

00:06:46 Speaker 0: Okay. And Also need to confirm your billing address?

00:06:50 Speaker 1: My address is 998 Lock Drive, in C Cove Main. The Zip code there is 11002K.

00:07:04 Speaker 0: You spell Lot?

00:07:06 Speaker 1: It's L0CKW00DE. Sorry. It should've have said there was an e on the end of that.

00:07:15 Speaker 0: No. That's perfectly fine. Alright. So I have down that I am canceling your plan right now. You should start receiving charges billing cycle.

00:07:25 Speaker 0: If for some reason, you continue to receive charges, this contact support who will fix this problem for you.

00:07:31 Speaker 1: Alright. Thanks for your help.

00:07:33 Speaker 0: Not a problem. Is there anything else I can I help you with?

00:07:36 Speaker 1: No. That's all I needed.

00:07:37 Speaker 0: Alright. Well, thank you for calling, And I hope you have a nice day.

00:07:40 Speaker 1: Yep. You too. Bye. Bye.
```

**Comment by CarmAero-1** [comment]
This is fantastic!

 

Thanks for your help and suggestions, and I must say that you represent your company very well. Thank you for the time and effort you put into assisting us with the code. It's exactly what we were looking for. 

 

We will review the AI-powered speech analytics information you provided, and I believe it may meet our needs. Your company's innovative and prudent approach to this subject matter is comprehensive and well-thought-out. 

 

As the System Architect on this project, I will need some time to process this unexpectedly positive turn of events, as it deviates from our usual piecemeal approach. 

 

Once I finish building the proof of concept, and if it is successful, I will present it to my team and we will be migrating to your company's offering.

Best regards,
[redacted]
